### PR TITLE
Support copying files in volume

### PR DESCRIPTION
**Pull request by fahedouch**
fixing  https://github.com/containerd/nerdctl/issues/165

### CODE REVIEW

File: run_mount.go [hunk 1]
```diff
@@ -69,12 +77,59 @@ func generateMountOpts(clicontext *cli.Context, imageVolumes map[string]struct{}
...
+
+		diffIDs, err := i.RootFS(ctx)
+		if err != nil {
+			return nil, nil, err
+		}
+		chainID := identity.ChainID(diffIDs).String()
+
+		s := client.SnapshotService(clicontext.String("snapshotter"))
+
+		tempDir, err := ioutil.TempDir(os.Getenv("GOTMPDIR"), "initialC")
```
Comment (AkihiroSuda) [inline run_mount.go:113]: Why GOTMPDIR

File: run_mount.go [hunk 2]
```diff
@@ -69,12 +77,59 @@ func generateMountOpts(clicontext *cli.Context, imageVolumes map[string]struct{}
...
+		if err := mount.All(mounts, tempDir); err != nil {
+			if err := s.Remove(ctx, tempDir); err != nil && !errdefs.IsNotFound(err) {
+				return nil, nil, err
+			}
+			return nil, nil, err
+		}
+		target := strings.Join([]string{tempDir, imgVol}, "")
+
+		//copying up initial contents of the mount point directory
+		copyExistingContents(target, anonVol.Mountpoint)
```
Comment (AkihiroSuda) [inline run_mount.go:131]: Named volumes need this as well

File: run_mount.go (hunk 1 above)
Comment (fahedouch) [reply run_mount.go:113]: it is juste a choice. I can create tempDir in `.` it doesn't change anything for me

Comment (fahedouch) [reply run_mount.go]: updated to `tempDir, err := ioutil.TempDir("", "initialC")`

File: run_mount.go (hunk 2 above)
Comment (fahedouch) [reply run_mount.go]: @AkihiroSuda  I made modification. FYI : in docker named volume do not copy initial content ( that may override source volume)

File: run_mount_test.go [hunk 3]
```diff
@@ -98,3 +98,29 @@ VOLUME /foo
...
+	assert.NilError(t, err)
+	defer os.RemoveAll(buildCtx)
+
+	base.Cmd("build", "-t", imageName, buildCtx).AssertOK()
+	//AnonymousVolume
+	base.Cmd("run", "--rm", imageName).AssertOut("hi")
+	base.Cmd("run", "-v", "/mnt", "--rm", imageName).AssertOut("hi")
+
+	//NamedVolume
+	base.Cmd("run", "-v", "/home:/mnt", "--rm", imageName).AssertFail()
```
Comment (AkihiroSuda) [inline run_mount_test.go:125]: This is a bind mount, not a named volume (nerdctl volume create)

File: run_mount.go [hunk 4]
```diff
@@ -51,6 +100,19 @@ func generateMountOpts(clicontext *cli.Context, imageVolumes map[string]struct{}
 			}
 			ociMounts[i] = x.Mount
 			mounted[filepath.Clean(x.Mount.Destination)] = struct{}{}
+
+			//copying up initial contents of the mount point directory
+			for imgVolRaw := range imageVolumes {
+				imgVol := filepath.Clean(imgVolRaw)
+				target := strings.Join([]string{tempDir, imgVol}, "")
+
+				//Coyping content only for AnonymousVolume
```
Comment (AkihiroSuda) [inline run_mount.go:109]: No

File: run_mount.go [hunk 5]
```diff
@@ -41,6 +49,47 @@ func generateMountOpts(clicontext *cli.Context, imageVolumes map[string]struct{}
 		anonVolumes []string
 	)
 	mounted := make(map[string]struct{})
+	var imageVolumes map[string]struct{}
+	var tempDir string
+
+	if ensuredImage != nil {
+		imageVolumes = ensuredImage.ImageConfig.Volumes
+		client, ctx, cancel, err := newClient(clicontext)
```
Comment (AkihiroSuda) [inline run_mount.go:57]: No, we already have the client

**Review by AkihiroSuda — CHANGES_REQUESTED** [review]

File: run_mount.go (hunk 5 above)
Comment (fahedouch) [reply run_mount.go:57]: fixed

File: run_mount.go (hunk 4 above)
Comment (fahedouch) [reply run_mount.go:109]: fixed

File: run_mount.go (hunk 2 above)
Comment (fahedouch) [reply run_mount.go]: @AkihiroSuda sorry I mixed mount bind and named volume. I added `copyExistingContents`to named volume

File: pkg/mountutil/mountutil.go [hunk 6]
```diff
@@ -90,7 +96,7 @@ func ProcessFlagV(s string, volStore volumestore.VolumeStore) (*Processed, error
 		return nil, errors.Errorf("failed to parse %q", s)
 	}
 	res.Mount = specs.Mount{
-		Type:        "none",
+		Type:        string(vType),
```
Comment (AkihiroSuda) [inline pkg/mountutil/mountutil.go:99]: This is an OCI Mount object, you can’t mix up Docker API type here

File: run_mount.go (hunk 5 above)
Comment (AkihiroSuda) [reply run_mount.go:59]: Doesn’t seem fixed

File: run_mount_test.go [hunk 7]
```diff
@@ -98,3 +98,33 @@ VOLUME /foo
...
+	base.Cmd("build", "-t", imageName, buildCtx).AssertOK()
+	//AnonymousVolume
+	base.Cmd("run", "--rm", imageName).AssertOut("hi")
+	base.Cmd("run", "-v", "/mnt", "--rm", imageName).AssertOut("hi")
+
+    //NamedVolume
+	base.Cmd("volume","create", "copying-initial-content").AssertOK()
+	base.Cmd("run", "-v", "copying-initial-content:/mnt", "--rm", imageName).AssertOut("hi")
+	//mount bind
+	base.Cmd("run", "-v", "/home:/mnt", "--rm", imageName).AssertFail()
```
Comment (AkihiroSuda) [inline run_mount_test.go:129]: Mounting /home is dangerous, please use some inutile.TempDir

File: pkg/mountutil/mountutil.go (hunk 6 above)
Comment (fahedouch) [reply pkg/mountutil/mountutil.go:99]: @AkihiroSuda does OCI provided types ? or I may set the directly in the code ?

Comment (fahedouch) [reply pkg/mountutil/mountutil.go]: @AkihiroSuda  is it ok for you ?

File: run_mount.go (hunk 5 above)
Comment (fahedouch) [reply run_mount.go]: it should be ok now!

File: pkg/mountutil/mountutil.go [hunk 8]
```diff
@@ -90,7 +93,7 @@ func ProcessFlagV(s string, volStore volumestore.VolumeStore) (*Processed, error
 		return nil, errors.Errorf("failed to parse %q", s)
 	}
 	res.Mount = specs.Mount{
-		Type:        "none",
+		Type:        vType,
```
Comment (AkihiroSuda) [inline pkg/mountutil/mountutil.go:96]: Again, you can't have "volume" "bind" here

https://github.com/opencontainers/runtime-spec/blob/master/config.md#posix-platform-mounts

**Review by AkihiroSuda — CHANGES_REQUESTED** [review]

File: pkg/mountutil/mountutil.go (hunk 8 above)
Comment (AkihiroSuda) [reply pkg/mountutil/mountutil.go:96]: The type fIeld can be  in this struct https://github.com/containerd/nerdctl/blob/86e8f5f10d46fc51fb152733a025d940f91ae668/pkg/mountutil/mountutil.go#L32

Comment (fahedouch) [reply pkg/mountutil/mountutil.go]: @AkihiroSuda moved.

File: run_mount_test.go [hunk 9]
```diff
@@ -98,3 +99,39 @@ VOLUME /foo
...
+	//NamedVolume
+	base.Cmd("volume", "create", "copying-initial-content").AssertOK()
+	base.Cmd("run", "-v", "copying-initial-content:/mnt", "--rm", imageName).AssertOutContains("hi")
+
+	//mount bind
+	tmpDir, err := ioutil.TempDir("", "hostDir")
+	assert.NilError(t, err)
+	defer os.RemoveAll(tmpDir)
+
+	m := strings.Join([]string{"./", tmpDir, ":", "/mnt"}, "")
```
Comment (AkihiroSuda) [inline run_mount_test.go:135]: You don’t need ./ here

File: pkg/mountutil/mountutil.go [hunk 10]
```diff
@@ -32,6 +32,7 @@ import (
 type Processed struct {
 	Mount           specs.Mount
 	AnonymousVolume string // name
+	Type            string
```
Comment (AkihiroSuda) [inline pkg/mountutil/mountutil.go:35]: enum (Bind = iota, Volume)

**Comment by fahedouch** [comment]
@AkihiroSuda  any idea what should disturb the `cgroup2` job ?

File: run_mount.go [hunk 11]
```diff
@@ -41,6 +51,42 @@ func generateMountOpts(clicontext *cli.Context, imageVolumes map[string]struct{}
...
+		}
+		defer os.RemoveAll(tempDir)
+
+		var mounts []mount.Mount
+		mounts, err = s.View(ctx, tempDir, chainID)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		if err := mount.All(mounts, tempDir); err != nil {
```
Comment (dmcgowan) [inline run_mount.go:83]: Where are the mounts getting cleaned up? The temp mount logic is ideal for these type of use cases

Comment (fahedouch) [reply run_mount.go:88]: @dmcgowan I added unmount logic .  The temp mount logic is  ideal but not in this context because it  push me to make same steps ( tmp dir creation then mount then unmount ) in each iteration `flagVSlice`

File: pkg/mountutil/mountutil.go (hunk 10 above)
Comment (fahedouch) [reply pkg/mountutil/mountutil.go:40]: @AkihiroSuda can review plz. is it ok ?

File: run_mount_test.go (hunk 9 above)
Comment (fahedouch) [reply run_mount_test.go]: fixed

**Review by AkihiroSuda — APPROVED** [review]
Pushed a new test.

LGTM if green. Thanks and sorry for slow reviewing

File: run_mount.go [hunk 12]
```diff
@@ -51,6 +105,19 @@ func generateMountOpts(clicontext *cli.Context, imageVolumes map[string]struct{}
 			}
 			ociMounts[i] = x.Mount
 			mounted[filepath.Clean(x.Mount.Destination)] = struct{}{}
+
+			//copying up initial contents of the mount point directory
+			for imgVolRaw := range imageVolumes {
+				imgVol := filepath.Clean(imgVolRaw)
+				target := strings.Join([]string{tempDir, imgVol}, "")
```
Comment (ktock) [inline run_mount.go:112]: nit: why not `filepath.Join`?

File: run_mount.go [hunk 13]
```diff
@@ -69,12 +136,20 @@ func generateMountOpts(clicontext *cli.Context, imageVolumes map[string]struct{}
...
 		anonVolName := idgen.GenerateID()
+
 		logrus.Debugf("creating anonymous volume %q, for \"VOLUME %s\"",
 			anonVolName, imgVolRaw)
 		anonVol, err := volStore.Create(anonVolName)
 		if err != nil {
 			return nil, nil, err
 		}
+
+		target := strings.Join([]string{tempDir, imgVol}, "")
```
Comment (ktock) [inline run_mount.go:147]: ditto

**Comment by ktock** [comment]
needs rebase

```
Conflicting files
go.mod
```

File: run_mount.go (hunk 12 above)
Comment (AkihiroSuda) [reply run_mount.go:112]: we should use securejoin. I'll push a commit

File: run_mount.go (hunk 13 above)
Comment (AkihiroSuda) [reply run_mount.go]: done

**Review by ktock — APPROVED** [review]

**Comment by fahedouch** [comment]
@AkihiroSuda  should we test hardlink too ?

File: run_mount.go (hunk 11 above)
Comment (fuweid) [reply run_mount.go:91]: I think it depends on `user cannot CONTRL-C during copy`. Need to figure out how to clean up in this case

File: run_mount.go [hunk 14]
```diff
@@ -41,6 +51,50 @@ func generateMountOpts(clicontext *cli.Context, imageVolumes map[string]struct{}
...
+			return nil, nil, err
+		}
+
+		// We should do defer first, if not we will not do Unmount when only a part of Mounts are failed.
+		defer func() {
+			err = mount.UnmountAll(tempDir, 0)
+		}()
+
+		if err := mount.All(mounts, tempDir); err != nil {
+			if err := s.Remove(ctx, tempDir); err != nil && !errdefs.IsNotFound(err) {
```
Comment (fuweid) [inline run_mount.go:92]: need to umount first and then remove

**Comment by fuweid** [comment]
Using `tar` and `echo` might be safety, like what `kubectl cp`  does. It is hard to handle corner cases in `nerdctl` side.

WDYT? @AkihiroSuda @dmcgowan

**Comment by AkihiroSuda** [comment]
@fahedouch 

> should we test hardlink too ?

Yes, that's more preferrable.

- - -
@fuweid

> Using tar and echo might be safety, like what kubectl cp does. It is hard to handle corner cases in nerdctl side.

How do you use `echo`?
Could you open an issue/PR?

File: run_mount.go (hunk 11 above)
Comment (fahedouch) [reply run_mount.go:91]: https://github.com/containerd/nerdctl/pull/216

File: run_mount.go (hunk 14 above)
Comment (fahedouch) [reply run_mount.go:92]: https://github.com/containerd/nerdctl/pull/216
